File: branches/experimental/src/Gui/Graphics/PatternLegendItem.py
from LegendItem import *

from BrushFactory import *

class PatternLegendItem(LegendItem):
    def __init__(self, parent, scene, rect, pen, font, dataset):
        LegendItem.__init__(self, parent, scene, rect, pen, font, dataset)
        self.drawDatasetPattern()
        self.drawDatasetNumber()
        self.drawDatasetName()
    def drawDatasetPattern(self):
        w = self.rect().width() * 0.8
        self.patternItm = QGraphicsRectItem(self, self.scene())
        self.patternItm.setRect(QRectF(0, 0, w, w))
        self.patternItm.setPen(Qt.black)
        if self.dataset.hasSvgItem():
            f = BrushFactory()
            brush = f.fromSvgItem(self.dataset.svgItem)
            self.patternItm.setBrush(brush)

        self.patternItm.setPos(QPointF(self.rect().width() * 0.2 / 2,
                                       self.rect().width() * 0.2 / 2))
    def drawDatasetNumber(self):
        self.idItm = QGraphicsTextItem(QString("%1").arg(self.dataset.id), self, self.scene())
        self.idItm.setFont(self.font)
        self.idItm.adjustSize()
        self.idItm.setPos(QPointF((self.rect().width() - self.idItm.boundingRect().width()) / 2.0,
                                  self.patternItm.pos().y() + self.patternItm.rect().height()))
    def drawDatasetName(self):
        self.nameItm = QGraphicsTextItem(self.dataset.name.replace(' ', '\n'), self, self.scene())
        self.nameItm.setFont(self.font)
        self.nameItm.adjustSize()
        self.nameItm.setPos(QPointF((self.rect().width() - self.nameItm.boundingRect().width()) / 2.0,
                                    self.idItm.pos().y() + self.idItm.boundingRect().height()))

File: branches/cpp-impl/src/columnView/PatternLegendItem.cpp
/* 
 * File:   PatternLegendItem.cpp
 * Author: jolo
 * 
 * Created on 14. Januar 2010, 09:30
 */

#include "PatternLegendItem.h"

#include <QPen>

PatternLegendItem::PatternLegendItem(QGraphicsItem* p, int w, int h)
: LegendItem(p, w, h) {
}

PatternLegendItem::~PatternLegendItem() {
}

void PatternLegendItem::addPattern(const QBrush& b) {
    QPen p;

    p.setColor(Qt::black);
    p.setStyle(Qt::SolidLine);

    QGraphicsRectItem* pattern = new QGraphicsRectItem(this);
    pattern->setBrush(b);
    pattern->setPen(p);

    pattern->setRect(QRectF(0, 0, rect().width() * 0.8, rect().width() * 0.8));
    _space = (int)((rect().width() - pattern->rect().width()) / 2);
    pattern->setPos(_space, _space);

    _currHeight = (int)(_space + pattern->rect().height());
}
